### PR TITLE
Clean up of Python modernize

### LINKED ISSUE #485: Some uses of cmp still in codebase
The use of `cmp` isn't fixed by 2to3 or any other automatic tool.  For Python 3 compatibility these need to be manually changed (presumably to `key` functions).

Although these are in WxPython code, so not currently being run under Python3:
```
traitsui/wx/image_slice.py:283:39: F821 undefined name 'cmp'
traitsui/wx/image_slice.py:312:39: F821 undefined name 'cmp'
```

### PR DESCRIPTION
**Pull request by corranwebster**
Automated updating of code is rarely perfect.  This PR makes some manual improvements which reduce the use of six and/or simplify code.

Particular changes made:

- don't worry about `range` list creation with range under Python 2 in most cases.
- remove some old uses of `cmp`.  Fixes #485
- clean up `__future__` imports
- use list comprehensions instead of `list(map(...))` where it makes sense
- clean up a few uses of `six.string_types` and similar
- avoid `io.open` for binary reads (and use binary reads for code parsing/exec)
- use `next(...value())` instead `list(...value())[0]`
- use `fn.__name__` instead of `fn.__code__.co_name`

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -52,8 +51,8 @@
 d = {}
 try:
     version_path = os.path.join(base_path, '..', '..', 'traitsui', '_version.py')
-    with open(version_path, 'r', encoding='utf8') as fp:
-        exec(fp.read(), d)
+    with open(version_path, 'rb') as fp:
```
Comment (mdickinson) [inline docs/source/conf.py]: Nitpick: reading a text file as bytes seems ... wrong. I think I prefer the original here, even with the ugliness of the extra  import.

File: traitsui/extras/demo.py [hunk 2]
```diff
@@ -160,7 +160,7 @@ def init(self, info):
         locals['__file__'] = df.path
         sys.modules['__main__'].__file__ = df.path
         try:
-            exec(compile(open(df.path).read(), df.path, 'exec'), locals, locals)
+            exec(compile(open(df.path, 'rb').read(), df.path, 'exec'), locals, locals)
```
Comment (mdickinson) [inline traitsui/extras/demo.py]: Would be nice to add the `with` wrappers here, too.

File: traitsui/qt4/list_editor.py [hunk 3]
```diff
@@ -34,8 +34,7 @@
 from .editor import Editor
 from .helper import IconButton
 from .menu import MakeMenu
-from six.moves import range
-
+\
```
Comment (mdickinson) [inline traitsui/qt4/list_editor.py]: Typo?

File: traitsui/ui.py [hunk 4]
```diff
@@ -352,7 +352,7 @@ def find(self, include):
 
         # Get the context 'object' (if available):
         if len(context) == 1:
-            object = list(context.values())[0]
+            object = next(context.values())
```
Comment (mdickinson) [inline traitsui/ui.py]: This looks wrong to me; does this work? I'm assuming that `context` is a `dict`, in which case `context.values()` isn't an iterator in either Python 2 or Python 3. `object, = context.values()` would work, if you can put up with that kind of "grit-on-Tim's-monitor" syntax. Or just leave the original.

(Same comment again below for `next(context.keys())`.)

File: traitsui/undo.py [hunk 5]
```diff
@@ -27,17 +27,18 @@
...
+
 
 #-------------------------------------------------------------------------
 #  Constants:
 #-------------------------------------------------------------------------
 
-NumericTypes = (int, int, float, complex)
-SimpleTypes = (str, six.text_type, int, int, float, complex)
+NumericTypes = six.integer_types + (float, complex)
+SimpleTypes = six.string_types + NumericTypes + (bytes,)
```
Comment (mdickinson) [inline traitsui/undo.py]: Could replace `six.string_types` with `(six.text_type,)` here to avoid including `str` twice on Python 2.

Comment (mdickinson) [reply traitsui/undo.py]: Though I'm not convinced that `bytes` deserves to be called a simple type in Python 3.

File: traitsui/wx/dnd_editor.py [hunk 6]
```diff
@@ -223,15 +222,11 @@ def _get_drag_data(self, data):
...
-            object = load(fh)
-        except:
-            object = None
-
-        if fh is not None:
-            fh.close()
+        with open(file_name, 'rb') as fh:
+            try:
+                object = load(fh)
+            except Exception:
```
Comment (mdickinson) [inline traitsui/wx/dnd_editor.py]: I want to suggest a more targeted `Exception` type here, but apparently `pickle.load` can produce a variety of exceptions (including `EOFError` and `UnpicklingError`), and I don't know what a sensible full list is.

File: traitsui/wx/image_slice.py [hunk 7]
```diff
@@ -280,7 +280,7 @@ def _analyze_bitmap(self):
             else:
                 matches = [(dy / 2, 1)]
         elif n > self.stretch_rows:
-            matches.sort(lambda l, r: cmp(r[1], l[1]))
+            matches.sort(key=lambda x: x[1])
```
Comment (mdickinson) [inline traitsui/wx/image_slice.py]: It looks to me as though there's a reversal going on in the original code. Should we be using `reverse=True` or some such in this sorting operation? (Same comment below.)

File: traitsui/qt4/list_editor.py (hunk 3 above)
Comment (corranwebster) [reply traitsui/qt4/list_editor.py]: Yep - UK keyboard with `\` where the return key rightfully should be.  Fortunately harmless in this instance.

File: traitsui/undo.py (hunk 5 above)
Comment (corranwebster) [reply traitsui/undo.py]: This is for the purposes of whether to merge undo items - looking at where it is used it comes into play if there is a sequence of things which has been modified, and then the same item of the same sequence gets modified immediately after, then it merges the two.  Not actually sure of the validity of this as a strategy, but it seems if strings are treated this way, so should bytes.

File: traitsui/wx/image_slice.py (hunk 7 above)
Comment (corranwebster) [reply traitsui/wx/image_slice.py]: This whole section of code has the feel that it should either be re-written to be vectorized, or it should be cythonized.  Python for loops in analysis of images is Wrong.

Comment (mdickinson) [reply traitsui/wx/image_slice.py]: This is in image analysis? 🙀 I didn't look further than the lines in front of me.

File: traitsui/ui.py (hunk 4 above)
Comment (mdickinson) [reply traitsui/ui.py]: `[object] = context.values()` is Yet Another Way. I'm told that some people like it.

**Pull request merged**